Data Layer: refactor dispatchRequest to avoid repetitive code

jsnajdr opened the pull request:
This PR introduces a new `handleDataRequest` API for Data Layer that's intended to replace the `dispatchRequest` function. By applying several simple refactorings to `dispatchRequest`, I think I made the function more usable and less verbose.

I applied the new API to the Simple Payments code (I worked on it recently) and it made the code much shorter (260 lines to 210) and I believe it's more readable.

What are the refactorings I did?

**Don't distinguish between mandatory and optional arguments**
All parameters for the handler are now part of one big object:
```js
handleDataRequest( {
  initiate,
  onSuccess,
  onProgress,
} );
```
In the past, the opinion about what's optional and what's not has changed quite often, each time requiring a refactoring of all call sites. Now, everything is a named parameter of a single `options` object. If a mandatory parameter is missing, I issue a console warning: experience shows that Calypso contributors tend to be annoyed by warnings and fix them quickly.

**Change the handler signature to a classic action creator**
Until now, the `initiate`/`onSuccess`/... handlers had an unusual signature like this:
```js
onSuccess = ( store, action, data ) => {
  store.dispatch( someOtherAction( action.siteId, data ) );
}
```
It's somewhat similar to a middleware, but it's not a middleware. It's something very data-layer-specific.

I'm changing this into a classic action creator:
```js
onSuccess = ( action, data ) => someOtherAction( action.siteId, data );
```
and do the `dispatch` call at one place, in the framework code in `handleDataRequest`. This removes a lot of verbosity from the code.

If your handler did something more complex, it can always return an action thunk:
```js
onSuccess = ( action, data ) => dispatch => {
  doSomething();
  dispatch( someOtherAction() );
  doSomethingElse();
}
```
But most handlers don't require a thunk at all. This change removes some mystery from the handlers and converts them into a familiar language of action creators and thunks, something every Redux developer is already fluent in.

Doing the dispatch in the framework code has another advantage: the framework can change the action before dispatching it. For example, `handleDataRequest` decorates the `http` actions with `onSuccess` and `onError` properties. Again, user code can be simplified and complexity is moved to the framework.

**Return array of handlers**
The `handleDataRequest` returns an one-element array of handler functions, so the return can be conveniently passed directly as part of `mergeHandlers` argument.

I personally think that supporting multiple handlers for a data-layer action is not a great idea. One handler is sufficient for a vast majority of cases and the rest can be converted. But most importantly, having one handler allows to return a value from the `dispatch` call. That'd be extremely useful for sequencing requests together (using Promises or other async primitives) and using them outside Redux (like in `page` handlers or React components with local state).

If we decide to do this change, the user code that uses `handleDataRequest` won't need to be changed -- only the framework internals will be updated.

**Example:**
This is an example usage of the new API:
```js
export const handleProductListEdit = handleDataRequest( {
	initiate: action =>
		http( {
			method: 'POST',
			path: `/sites/${ action.siteId }/posts/${ action.productId }`,
			body: productToCustomPost( action.product ),
		} ),
	fromApi: customPostToProduct,
	onSuccess: ( action, newProduct ) => receiveUpdateProduct( action.siteId, newProduct ),
} );

export default {
	[ SIMPLE_PAYMENTS_PRODUCTS_LIST_EDIT ]: handleProductListEdit,
};
```

**Further development:**
Here are some ideas on how to further improve the API:
1. Introduce nice API for creating the HTTP actions. The idea is to call `initiateAction = wpcom.site( siteId ).addPost( post )` -- the same idiomatic API we already know, but it just creates actions instead of executing the request. No need to know the details of the WP.com REST API (methods, paths).
2. Auto-generate the code that reads and updates the Redux state. Avoid creating the same CRUD actions, reducers and selectors over and over. Provide parametrized helpers that generate them.

nb commented:
Any reason we’re not removing `dispatchRequest` completely?

coderkevin commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -158,3 +163,73 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options ) => ( s
+	const data = getData( action );
+	if ( data ) {
+		try {
+			return onSuccess( action, fromApi( data ) );
```
Does this mean users of this code should throw an exception on invalid data rather than dispatching `onError` like we do now?

jsnajdr replied:
This `try/catch` block merely catches errors thrown by the `fromApi` transformer. Even if the network request is successful and the `onSuccess` handler is called, the final result can still be error.

If the network request fails, calling `onError` is still the right thing. There is no change from `dispatchRequest` here.

coderkevin commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -158,3 +163,73 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options ) => ( s
+	}
+
+	const initiator = initiate( action );
+	if ( initiator ) {
```
So, the desired implicit behavior if `initiate( action )` returns a falsy value is to just silently do nothing? That seems strange to me if someone accidentally passes a function that doesn't have a return or returns null.

jsnajdr replied:
Oh, that's a good catch! I pushed a commit where a warning is issued when `initiator` doesn't return anything.

coderkevin commented:
> Any reason we’re not removing dispatchRequest completely?

There's a lot of code out there using it still.

coderkevin commented:
> Introduce nice API for creating the HTTP actions. The idea is to call initiateAction = wpcom.site( siteId ).addPost( post ) -- the same idiomatic API we already know, but it just creates actions instead of executing the request. No need to know the details of the WP.com REST API (methods, paths).

I'd like to see the creation of API modules on top of this. For example, I'd like to create a WooCommerce REST API module that has action creators to be used by application code, then a middleware to handle them and make the appropriate HTTP request. This would be a natural abstraction point for hybrid extensions that could work both in wpcom using the wpcom-proxy and jetpack auth, and also in wp-admin using direct http calls and cookie auth.

dmsnell commented:
Real quick: this is a review without digging deep, so take it at that…

I'm wondering if this is _that much better_ than `dispatchRequest()` - in the example given I'm not sure 210 lines is much fewer than 260 in the scheme of things. Also I really don't want to introduce thunks here again do to several of the ways they hide state and break down some of the Redux norms.

I see a few things that reduce the amount of syntax we're writing (and the thunk seems to be a workaround around that) but beyond that I don't see much that's different about `handleDataRequest`

The returned arrays may not be necessary if all we're doing is dumping API data into Calypso, but they are extremely useful for when we are actually taking advantage of related data and events as Redux is designed to do.

I'm quite interested in making a radical departure from `dispatchRequest` for things that can significantly reduce the amount of shipped code and developer complexity: the hastily-named `apiDump()` that I know @samouri has also been interested in. did you have thoughts on the API I proposed in p2-p4TIVU-86D? (since then I've made one change in my mind to require the schema) in that case this whole system turns into about five lines per API resource - it's essentially a lightweight generic REST client

samouri commented:
I'm a fan of the function signatures proposed here.  It is [fairly similar](https://github.com/Automattic/wp-calypso/pull/12616) to a pr I made last March.  The format is both more concise and a bit more ergonomic than what we are currently doing with `dispatchRequest`.

>Change the handler signature to a classic action creator....

Right now there are some (read: at least one) data layer handlers that relies on having access to the store's state.   I know that comments code for loading more comments checks to see which is the oldest comment in the store and requests for any comments before that date. That said, it shouldn't be written the way it is and that logic should exist within the calling code.

---
@dmsnell re:
> Also I really don't want to introduce thunks here again do to several of the ways they hide state and break down some of the Redux norms

I don't agree with that. functions on their own aren't a bad thing, whats important is that side effects be conveyed via actions.  The only difference here with thunks vs. action creator is who gets control over `dispatch`.

---

I really like `handleDataRequest`, but I have a fear that we'd never port over the code currently using `dispatchRequest`.   I don't enjoy the idea of having multiple same-usecase patterns in the data-layer but I guess its inevitable at some point.  My question then: is `handleDataRequest` compelling enough to warrant the split?  Maybe it is! but I wanted to bring this question up.

> Further development:
> Auto-generate the code that reads and updates the Redux state. Avoid creating the same CRUD actions, reducers and selectors over and over. Provide parametrized helpers that generate them.

This bit is exciting :).  It seems to me like it could be added to either `dispatchRequest` or `makeDataRequest` equally easily.

jsnajdr commented:
Thanks everyone for your comments so far! I've been sick for the last few days and I'm still not 100% back at work, so there will be some delay before I can incorporate and respond to the feedback. Thanks again and stay tuned 😜

jsnajdr commented:
> I'm wondering if this is that much better than dispatchRequest() - in the example given I'm not sure 210 lines is much fewer than 260 in the scheme of things.

I'm a fan of doing incremental focused improvements that compound to a big change over time -- so don't expect any revolutionary PRs from me :) This change removes a lot of code from almost every Data Layer module and opens some new possibilities. And I believe that existing code could be codemodded, bringing migration costs close to zero. For a single PR, I'm quite satistifed with the result :)

> Also I really don't want to introduce thunks here again due to several of the ways they hide state and break down some of the Redux norms.

Every data layer handler is a thunk now, only slightly obfuscated. Change the handler signature from:
```js
( { dispatch, getState }, action, data ) => { ... }
```
to:
```js
( action, data ) => ( dispatch, getState ) => { ... }
```
and the offender is immediately uncovered!

The good news the any thunk non-lovers among us is that a lot of the thunkiness is going away -- there are many trivial thunks that can be transformed into mere action creators:
```js
// both variables are of type "Dispatchable"
const thunk  = dispatch => dispatch( creator() );
const notThunk = creator();
```

jsnajdr commented:
> The returned arrays may not be necessary if all we're doing is dumping API data into Calypso, but they are extremely useful for when we are actually taking advantage of related data and events as Redux is designed to do.

Is there any good place in Calypso where the array of handlers is used to perform something useful? So far, I've seen only 1:1 mappings from action types to effect handlers. I'm genuinely interested -- does anything become impossible without the support for multiple handlers?

Single handlers would make it easier to return a value (request ID or a Promise) from a `dispatch` call -- something very useful for updating and creating resources.

jsnajdr commented:
> did you have thoughts on the API I proposed in p4TIVU-86D-p2 ?

I like that proposal in general and I think we have a shared goal here. This PR is implementing some things that could be very helpful to achieve it -- simplify the handlers, make them more "declarative" by moving the "execution" to the framework...

On the other hand, I think the proposal is trying to do too many things at once and mixes together things that could be addressed individually, step-by-step:
- simplifiying the handlers
- implementing a common "CRUD reducer" to avoid the same thing over and over
- adding a "lastUpdated" field to the state, and issue conditional requests like `getSites( { maxAge: 60 } )` -- very similar to HTTP caching, something you call "freshness".

For that "too many things at once" reason, I also have a hard time understanding #19060 right now -- with all its "apiStuffers" and "middlewarers"... I have a hunch that multiple smaller, orthogonal and focused changes could achieve the same thing. But that's mostly my personal preference of evolution over revolution. I wish we could move forward incrementally and iterate quickly by building on each other's work.

dmsnell commented:
> Every data layer handler is a thunk now, only slightly obfuscated.

This is not true though and the function signatures given is a misleading comparison. Thunks are completely opaque to our action stream; they leave a blank in the Redux history which sits there ambiguously.

While it's true that the action handlers may call `dispatch()` multiple times it's not true that they have the same operating characteristics of `redux-thunk` actions - just the same way we wouldn't call a function in `mapDispatchToProps` a _thunk_ action just because it calls `dispatch` multiple times. The data layer handlers are static devices which leaves transparent trail in the Redux history. If we want to understand _why_ an action was fired we can easily map the preceding `_REQUEST` or other trapped action types to the points where they are defined in the source code. We cannot work backwards from a `redux-thunk` action to source code.

> Is there any good place in Calypso where the array of handlers is used to perform something useful?

After looking around I didn't see anything, although I have suggested it before in order to keep certain logic separate from each other. I think that the combination of `dispatchRequest()` and `fromApi` may obviate some of those cases.

I'm increasingly feeling like the idea of taking advantage of interrelated data types and the efficiencies possible because of that just won't take off, hence my `apiData()` proposal.

> does anything become impossible without the support for multiple handlers?

Yes - one has to be aware first of whether some other file is watching for a given action because we wouldn't be able to have two independent handlers watching for the same one. Right now it appears like you are correct in finding that nothing _is_.

> simplify the handlers, make them more "declarative" by moving the "execution" to the framework…evolution over revolution. I wish we could move forward incrementally and iterate quickly by building on each other's work.

Yeah I think we could do this with the existing system which is part of what caught me off guard on this PR - it's a base-level change instead of iterating on the existing tool.

There are very few handlers which dispatch more than once (comments use it) though I think predict that as the ideas here sink in more we'll find more people doing multiple dispatches. On the other hand I like getting rid of `dispatch()` and find it cumbersome as well, which is why I proposed #19058.

Many handlers are using `getState()`; it seems reasonable that we could examine a change to the handler API like a mix between what's presented here and what exists…

```haskell
handler :: ( Action, State ) => [ Action ]
```

…this would let us continue to be able to present multiple dispatch, would remove some of the confusion around the timing of dispatched events, and remove the need to call `dispatch()` manually. It presents a progression of complexity from argument-less handlers, to those dependent on the calling action, to those dependent also on existing state.

> the proposal is trying to do too many things at once

It's serving a very different purpose than the generic handlers of the data layer. It's specifically addressing [shoving](https://github.com/Automattic/wp-calypso/blob/7c6acd66c209f86e96942ba6e2f1ae961f4bc3db/client/state/data-layer/wpcom/meta/sms-country-codes/index.js) [handlers](https://github.com/Automattic/wp-calypso/blob/7c6acd66c209f86e96942ba6e2f1ae961f4bc3db/client/state/data-layer/wpcom/timezones/index.js) which aren't getting much benefit from the data layer anyway (compared to how much code they are writing). The "CRUDy" operations they are doing can be extracted without adding any code and we can rip them out of the app altogether.

dmsnell commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -161,3 +166,77 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) =
+	};
+
+	// Return an one-element array of handlers
+	return [ handler ];
```
noted in my comment in the main thread, I think there is something in this we might be able to apply more generally to the handler API.

going back to the question: "is this different enough to warrant a new fundamental piece?" - could we instead update the data layer API itself with the ideas behind these changes without losing anything we want from the data layer?

dmsnell commented:
> hard time understanding #19060 right now -- with all its "apiStuffers" and "middlewarers"

right now these things are more incidental than anything else. the important bits to understand are the API details for React components and for defining new data types. the _stuffers_ and _shovers_ and _middlewarererers_ are my short-hand for "I don't have a name yet" and "I don't know how to build the provider for this API yet"

> I have a hunch that multiple smaller, orthogonal and focused changes could achieve the same thing

I'm not persuaded of this as `apiData()` is attempting to be a kind of off-the-shelf "I want to connect to data coming from an API" provider. It's itself a data layer handler built upon the framework of the handlers, a composition of the handler API and `dispatchRequest()`

What's most different there is that since we know so much about the kinds of operations we want to do with "shoved API data" we can leave all that code out Calypso for each type and classify all of these into their own bit.

Simplifying handlers would probably make it easier to shift into `apiData()` though if we want to, but I think that the kinds of handlers which `apiData()` would replace are already of the most basic code-mod-able kind

jsnajdr commented:
> This is not true though and the function signatures given is a misleading comparison. Thunks are completely opaque to our action stream; they leave a blank in the Redux history which sits there ambiguously.

Comparing the two ways how to write a simple handler:
```js
handler1 = ( store, action, data ) => (
  store.dispatch( { type: 'UPDATE', siteId: action.siteId, data } )
);

handler2 = ( action, data ) => dispatch => (
  dispatch( { type: 'UPDATE', siteId: action.siteId, data } )
);
```
I don't see any difference in the action logs they leave behind: there is always the "request" that's handled by the data layer, and then the `UPDATE` action. Completely identical. The inner working are slightly different, with one thunk being dispatched, but the outcome as observed by Redux Devtools or any action logging middleware or by the reducer itself is identical.

The upside of `handler2` is that it can be simplified in many cases by removing the `dispatch => dispatch( action )` bit. I fail to see any downside at all.

jsnajdr commented:
> Yeah I think we could do this with the existing system which is part of what caught me off guard on this PR - it's a base-level change instead of iterating on the existing tool.

I see it as a incremental improvement of the `dispatchRequest` API -- pass parameters in a more flexible way, get rid of repetitive `dispatch`, hide the "array of handlers" stuff...

With some effort, we could make this either backward compatible with existing `dispatchRequest`, or easily codemoddable. 

> going back to the question: "is this different enough to warrant a new fundamental piece?"

If the question is posed as "is this different enough from `dispatchRequest` to warrant the effort of rewriting all 50+ existing handlers by hand?" then the answer is probably NO.

But the questions we should be asking is:
1. Do we see this as a positive change? Is it a better way to write handlers?
2. Considering the migration costs, is it good enough to make the whole operation "profitable"?
3. Can we drive the migration costs down to zero by either being backward compatible, or by codemodding?

If the answer to #1 is yes, then let's get creative and constructive about issue #3.

jsnajdr commented:
```js
handler :: ( Action, State ) => [ Action ]
```
> …this would let us continue to be able to present multiple dispatch, would remove some of the confusion around the timing of dispatched events, and remove the need to call dispatch() manually.

This "return array of actions" is exactly what @samouri did in #12616, isn't it?

The data layer handlers could be divided into two large groups:
1. The ones that just dispatch one simple action. Arguably a large majority
2. The other ones that execute arbitrary code, with access to `store`, `originalAction` and `data`.

If we introduce ability to return an array of actions, we make the group #2 slightly smaller, but still far from empty -- the "arbitrary" code need will not go away. So, the question is: is the group of "array of actions" handlers big enough to warrant a special treatment?

"arbitrary code that has access to `dispatch` and `getState`" can be by definition said by one word: thunk.

An example of a data-layer handler that has complex logic is polling for automated transfer status: dispatching a `fetchAutomatedTransferStatus` action will poll the `transfer/status` REST endpoint every 3 seconds, update the Redux state, and will end the polling once the status is `ERROR` or `COMPLETE`. We'll always need an "arbitrary code" option for things like this.

dmsnell commented:
@jsnajdr how much work do you think it would take to update the existing handler API to return actions to dispatch instead of directly dispatching them? that seems like one incremental improvement we all might be able to get behind (cc: @samouri) - do we need `dispatch` there at all? probably not.

> poll the transfer/status REST endpoint every 3 seconds, update the Redux state, and will end the polling once the status is ERROR or COMPLETE. We'll always need an "arbitrary code" option for things like this.

Yes, though I think we can eliminate most of it or possibly all of it if we coordinate with the API endpoints. Data freshness can take care of basic polling and even in these value-based polls we can still handle that through appropriate use of `fromApi()` and component `render()` (for example, include the data need only if status isn't `ERROR` or `COMPLETE`)

samouri commented:
> We'll always need an "arbitrary code" option for things like this.

We have a need for that in Reader as well.  This doesn't scale, but we currently load _all_ of a users followed sites when they boot up Reader.  We have something called a FOLLOWS_SYNC that will set a syncInProgress flag and then keep requesting and `dispatching` more pages of results until it hits the end.

jsnajdr commented:
> how much work do you think it would take to update the existing handler API to return actions to dispatch instead of directly dispatching them?

Updating the handlers should be a pretty straighforward codemod. I'm bit more concerned about the tests -- updating all of them to pass again could be more messy. Let's try it out! :)

> Yes, though I think we can eliminate most of it or possibly all of it if we coordinate with the API endpoints.

Yes, my takeaway is that there is always some "messy but working very well" code around. And a well-designed API should support things like this. There's almost always some way to make it beautiful, but in many cases, there's little chance that it ever becomes a priority to "fix" it.

dmsnell commented:
> I grepped and we are currently at 74 usages. While that won't be fun to update, we can try our hand at codemods and at worst, do it by hand. We aren't at a place yet where we are too deep to iterate on our building blocks.

I've updated several things and this isn't a big pain to do by hand. In fact I figured it worked out faster than creating a codemod.

> I think yes. It takes out much of the incidental complexity and the difference is illustrated best by applying this to our simplest handlers.

Hm…I guess I really don't see it this way. I struggle to see how it's much different and why we can't just iterate on the data layer handler API and `dispatchRequest()`.

That is, if we propose making the data layer handler API return a list of actions to dispatch and take away `dispatch` I see it doing basically everything this PR is trying to do except doing so without introducing a new concept.

jsnajdr commented:
> Hm…I guess I really don't see it this way. I struggle to see how it's much different and why we can't just iterate on the data layer handler API and dispatchRequest().

I'm perfectly fine with keeping the `dispatchRequest` function and just refactoring it to match the `handleDataRequest` proposal.

The only reason I created a second handler from scratch was that I couldn't even create a working PR without it: existing code with 70+ handlers continues to work, and there is a small "experimental area" that presents the new API and uses it on a small sample of app code.

jsnajdr commented:
@samouri @dmsnell I just pushed an update with the following changes:
- instead of introducing a new `handleDataRequest` helper, I updated the existing `dispatchRequest`. It breaks all the existing code that uses it and is not updated yet, but the incremental improvement done here doesn't warrant a new name.
- updated the `README` that describes the `dispatchRequest` helper
- updated the tests for the helper to pass and to make the test code look good
- updated Simple Payments data layer code to use the new `dispatchRequest` API

The changes to `dispatchRequest` are:
- pass arguments as one big object with optional named params
- change the signature of handlers to action creators
- automatically decorate the initiator actions with `onSuccess/onError/onProgress` properties -- no need to pass `action` as a second param to `http` everywhere

I dropped the change that returned one-element array of handlers instead of the handler itself. It made updating the tests much more tedious and it's not really important at this moment. In other words, you still need to pass an array when defining the effect handler:
```js
export default {
  [ LIST_SITES ]: [ dispatchRequest( ... ) ]
}
```

If this looks good, can we start updating all the 74 `dispatchRequest` usages? Can we divide the work and coordinate?

dmsnell commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -110,54 +105,94 @@ const defaultOptions = {
+
+	return ( store, action ) => {
+		// create the low-level action we want to dispatch
+		const requestAction = createRequestAction( options, action );
```
is there a reason to split `createRequestAction` out into a separate function? It kind of seems like this function is doing nothing but validate the required parameters

jsnajdr replied:
`createRequestAction` returns an optional action that can be also `undefined` -- in case of a `noop` handler. Then we check if there is anything to dispatch and call `dispatch` if it is. This logic would be harder to express inside one big function.

It's almost always a good thing if a big function can be split into two or more smaller ones.

samouri commented on client/state/data-layer/wpcom-http/README.md:
```diff
@@ -179,16 +179,22 @@ dispatch( http( {
 
 // create request when no meta present, add on success, alert on failure
-dispatchRequest( fetchMenu, addMenuItems, alertFailure );
+dispatchRequest( { initiator: fetchMenu, onSuccess: addMenuItems, onError: alertFailure } );
```
can we change to `fetch: fetchMenu`?

jsnajdr replied:
Sure thing! `initiator` was renamed to `fetch` in latest push.

samouri commented on client/state/data-layer/wpcom/sites/simple-payments/index.js:
```diff
@@ -117,143 +140,69 @@ export function productToCustomPost( product ) {
-	[ SIMPLE_PAYMENTS_PRODUCTS_LIST_DELETE ]: [
-		dispatchRequest( requestSimplePaymentsProductDelete, deleteProduct, noop ),
-	],
+	[ SIMPLE_PAYMENTS_PRODUCT_GET ]: handleProductGet,
```
need to wrap all the values in square brackets now that dispatchRequest doesn't return an array. i.e.:

```js
export default {
  [ SIMPLE_PAYMENTS_PRODUCT_GET ]: [ handleProductGet ]  
}
```

jsnajdr replied:
Good catch! Fixed and tested 👍

samouri commented:
code lgtm.  do you think it would make more sense to leave the old dispatchRequest around temporarily as `dispatchRequestV0` and if the first arg isn't an object then redirect to it (all current handlers). That would obviate the need for one big scary pull request that updates everything at once and instead we can do a few folders at a time

dmsnell commented:
> do you think it would make more sense to leave the old dispatchRequest around temporarily as dispatchRequestV0 and if the first arg isn't an object then redirect to it (all current handlers)

or maybe even easier, just leave all existing `dispatchRequest()` calls and introduce the new one side-by-side as `dispatchRequest_v2()`, then move everything over as we can, then once `dispatchRequest()` is officially gone, we do one big rename from `_v2` to just `disaptchRequest`?

good idea though overall @samouri

dmsnell commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -110,54 +105,94 @@ const defaultOptions = {
+		onSuccess: action,
+		onError: action,
+		onProgress: action,
+	} );
```
do you think it would be good to only assign these to network-type requests? we currently have just two: `WPCOM_HTTP_REQUEST` and another raw HTTP request.

wouldn't see why normal actions need these, especially since they are going onto the top-level of the action

jsnajdr replied:
I'm not sure how to distinguish between network-type requests and others. I want to avoid enumerating some "special" action types inside `dispatchRequest` and give them special treatment.

I decided to roll back this change and start passing a second `action` parameter to the `http()` action creator again. The amount of typing it saved is not that great. `dispatchRequest` doesn't decorate the initiator action in any way and only dispatches it.

dmsnell commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -83,7 +88,7 @@ export const makeParser = ( schema, schemaOptions = {}, transformer = identity )
 			return transformer( data );
 		} catch ( e ) {
-			throw new TransformerError( e, transformer, data );
+			throw new TransformerError( e, data, transformer );
```
what's the swap of arguments here for?

jsnajdr replied:
I wanted to make the `transformer` parameter more optional. I started to throw `TransformerError` from a custom transformer (see `data-layer/wpcom/sites/simple-payments/index.js`) and the pointer to the transformer function is not always available. It can be an inline lambda function etc.

I'm also not sure if the `transformer` function pointer is useful at all -- can we extract any useful info to report from it?

dmsnell replied:
not sure - I don't know of any immediate need so let's rip it out - easy enough to add later if we need it

dmsnell commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -46,18 +51,18 @@ export const getHeaders = action => get( action, 'meta.dataLayer.headers', undef
 		this.inputData = data;
+		this.transformer = transformer;
 	}
 }
```
fun fact! I'm second-guessing my inclusion of a transformer error class. we can `throw` _anything_ from inside of there, which I think can be a valid way to signal different kinds of failures.

if it's a schema validation error we can check `instanceof SchemaError` or `isinstance NetworkError` but for the general `fromApi()` interface I can imagine common `throw`s of things like `false` or actual data, such as a some error object - things that exception classes may not be necessary for

(this is a comment not related to this PR)

jsnajdr commented:
@dmsnell @samouri I addressed a lot of review comments and renamed the new API to `dispatchRequestEx` (respecting the great naming tradition of Win32 APIs) while keeping the old version under the `dispatchRequest` name.

The PR is ready for another round of review and incremental transition after it's merged.

samouri commented on client/state/data-layer/wpcom-http/README.md:
````diff
@@ -179,16 +179,22 @@ dispatch( http( {
+} );
 ```
 
 ### `dispatchRequest()` helper
````
shall we temporarily renamed this to `dispatchRequestEx` as well?

jsnajdr replied:
I'm not sure what to do with docs. Include documentation for both variants now? Or only the new one and leave the documentation a bit confusing for a while?

Experience says that the upgrade won't happen overnight.

samouri replied:
> Experience says that the upgrade won't happen overnight.
 Write

having both variants in the docs makes sense, but that sounds like too much churn for my tastes. 

Personally I would prefer we just have the `Ex` variant for now (and not the docs for the old one).  Thats because all new code should use the new variant.  Any old code that is modified is most likely going to be touched by people that are already familiar with the old code.  You could even leave a link in the docs pointing to an older commit with the old docs as a temporary measure

dmsnell replied:
I'd be happy to help convert old usages. I'm fine keeping both in for a while, but if we think we can get all the updates made in one go _and_ catch in-progress PRs then would that make it easier? just do a clean update?

dmsnell commented on client/state/data-layer/wpcom-http/test/utils.js:
```diff
@@ -236,4 +243,144 @@ describe( 'WPCOM HTTP Data Layer', () => {
+				type: 'object',
+				properties: { count: { type: 'string' } },
+			} );
+			dispatchRequestEx( { fetch, onSuccess, onError, fromApi } )( store, successHttpAction );
```
some of these variables, particularly here, are making it hard for me to see what the test is doing - would there be any reason not to inline some of the actions/objects which are test-specific and contain the things the test is paying attention to?

for example, here I'm wondering about the `count` property - are we sending it in the test? having the object right here would help.

jsnajdr replied:
If you have ideas to improve the tests, please don't hesitate and push a commit please. I definitely agree there's a room for improvement.

dmsnell commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -161,3 +166,101 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) =
+		// create the low-level action we want to dispatch
+		const requestAction = createRequestAction( options, action );
+		// dispatch the low level action (if any was created) and return the result
+		return requestAction ? store.dispatch( requestAction ) : undefined;
```
Not a big deal here, but if we change the return value from `createRequestAction` to a list we can just dispatch each item and then the empty case will be handled automatically by means of the safety that empty lists already provide.

```js
processAction( options, action ).forEach( store.dispatch );
```

jsnajdr replied:
`dispatch` can return a value and I'd like to avoid adding places where it's thrown away. I have some ideas down the road that could greatly improve data layer ergonomics and these return values will be important there.

Using the array is a nice monadic trick to avoid a conditional, but I think it's an unnecessary complication -- you need to stop and think for a while to decode what it does.

dmsnell replied:
I'm just thinking of cases where we know we have arrays to return, plus the notion that the data layer doesn't return anything - it's all loosely coupled asynchronously

did you have something you know you want to return from dispatch and use somewhere?

jsnajdr replied:
I have many diverse use cases where returning a Promise from the dispatch would make complicated things simple and impossible things possible. Without violating any of the data layer design principles. But that's definitely for another PR -- we will surely have a vigorous debate about that soon 😉

jsnajdr commented:
I'm going to merge now. Next steps are:
- announce with a P2
- migrate all usages to `dispatchRequestEx` as quickly as possible
- rename `dispatchRequestEx` to `dispatchRequest`
- guard for old style arguments in `dispatchRequest` (function as first arg, second arg is not `undefined`...) and throw exception in that case. `dispatchRequest` is called statically when Calypso boots up, so all attempts to call the old version in in-progress PRs will fail very quickly and there is no danger of runtime exceptions.

gwwar commented:
@jsnajdr How are next steps looking? This looks like a great improvement, but the difference between `dispatchRequestEx` and `dispatchRequest` is quite hard to spot in code. I had to do a bit of spelunking to understand why certain handlers didn't match what I expected in terms of signatures. It'd be good to at least p2 that we have a new pattern for folks 😄

jsnajdr commented:
Thanks for the reminder @gwwar! I just published the long-overdue P2 :smile:
